### DISCUSSION TITLE
Discarding out _ carries ref dependencies.

### DISCUSSION THREAD

**Discussion by acaly** [dotnet/csharplang#5120] [Q&A] [answered]
When using `CollectionsMarshal.GetValueRefOrAddDefault`, I found the `out bool exists` blocks certain valid usage:

```c#
private readonly Dictionary<int, T> _myDict = new() { ... };

public static ref T Get(int i)
{
    return ref CollectionsMarshal.GetValueRefOrAddDefault(_myDict, i, out _); //Compiler error.
}
```

Ideally there should be an overload without the `out bool`, but I am not even allowed to write that overload myself. Is it possible that `out _` be resolved to something else to avoid the ref dependency? Currently what I can do is to use a static field:

```c#
private static bool _neverUsed;

public static ref T Get(int i)
{
    return ref CollectionsMarshal.GetValueRefOrAddDefault(_myDict, i, out _neverUsed); //ok
}
```

Since I think this is a language feature, can we have an attribute added to methods like this to indicate they don't pass the ref parameter to the returned ref and allow such methods to be called with `return ref ...(out _)`? Or do you think it's better to add an overload in the core library?

**Comment by 333fred** [comment, marked as answer]
https://github.com/dotnet/csharplang/blob/main/proposals/low-level-struct-improvements.md

See the parts about `DoesNotEscape`.

NB: In the future, it would help if you included the compiler error, rather than just saying "Compiler error". I had no idea what error it was and had to go create an example so I could see what you were talking about.

**Reply by acaly** [reply]
Thanks!

So it's CS8347, in case someone else wonders.

**Reply by 333fred** [reply]
https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4FgBQABADAAmwEYAWAbiy2wCZ8BhAHgBUA+fAbywAcAnKANwgAXAKaEiANnw8REBAHsYAGwCe+ACJQAxkKiKIPFQ1hC4+VvgD6YFZp34AvPhgBXJUoqYqAZnFSZAGbm+ADiIkIAFCb4UACUWJyY+MmEAOzSIkF08u4iOnowAM4AsgaFABYQSgB0YUIAalUuIgBKmQDyPACCCAjqmRBukda22qYxZvIuQlaxZPgA9AvZYFxQSiI8+Js88jzVWAC+R5Q4tNm5+YolZZVKHFgphL7E/pnmjUrNAPyh4Z/NNoBTo9PoDIbMADSIhUcCYAJEbAixBI1Tsun0hihMLhCLYCDGBQMsPM0LUAGscfgpjNgPIctsAB5QQpCQrxJIONhCcq7ADuzjcHiOQA=== was the example I made, for future reference as well.
